File: sdks/typescript/src/core/lifecycle.ts
/**
 * DCP-05 v2.0 Agent Lifecycle Management.
 *
 * Implements the agent lifecycle state machine:
 *   commissioned → active → declining → decommissioned
 *
 * Provides commissioning certificates, vitality reports (hash-chained),
 * decommissioning records, vitality scoring, and state transition validation.
 */

import { createHash } from 'crypto';
import type { CompositeSignature } from './composite-sig.js';
import type { CompositeKeyPair } from './composite-ops.js';
import { compositeSign } from './composite-ops.js';
import type { AlgorithmRegistry } from './crypto-registry.js';
import { canonicalizeV2 } from './canonicalize.js';
import { DCP_CONTEXTS } from './domain-separation.js';
import type {
  LifecycleState,
  TerminationMode,
  DataDisposition,
  VitalityMetrics,
  CommissioningCertificate,
  VitalityReport,
  DecommissioningRecord,
  Capability,
  RiskTier,
} from '../types/v2.js';

// ── Lifecycle State Machine ──

const VALID_TRANSITIONS: Record<LifecycleState, LifecycleState[]> = {
  commissioned: ['active', 'decommissioned'],
  active: ['declining', 'decommissioned'],
  declining: ['decommissioned', 'active'],
  decommissioned: [],
};

/**
 * Validate whether a state transition is allowed.
 */
export function validateStateTransition(from: LifecycleState, to: LifecycleState): boolean {
  return VALID_TRANSITIONS[from]?.includes(to) ?? false;
}

// ── Vitality Scoring ──

const METRIC_WEIGHTS = {
  task_completion_rate: 0.3,
  error_rate: 0.25,
  human_satisfaction: 0.25,
  policy_alignment: 0.2,
} as const;

/**
 * Compute a vitality score (0–1000) from metrics.
 * Higher is better. Error rate is inverted (lower error = higher score).
 */
export function computeVitalityScore(metrics: VitalityMetrics): number {
  const raw =
    metrics.task_completion_rate * METRIC_WEIGHTS.task_completion_rate +
    (1 - metrics.error_rate) * METRIC_WEIGHTS.error_rate +
    metrics.human_satisfaction * METRIC_WEIGHTS.human_satisfaction +
    metrics.policy_alignment * METRIC_WEIGHTS.policy_alignment;

  return Math.round(Math.max(0, Math.min(1, raw)) * 1000);
}

// ── Artifact Creation ──

/**
 * Create a commissioning certificate for a new agent.
 */
export async function createCommissioningCertificate(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    agent_id: string;
    session_nonce: string;
    human_id: string;
    commissioning_authority: string;
    purpose: string;
    initial_capabilities: Capability[];
    risk_tier: RiskTier;
    principal_binding_reference: string;
  },
): Promise<CommissioningCertificate> {
  const payload = {
    dcp_version: '2.0' as const,
    agent_id: params.agent_id,
    session_nonce: params.session_nonce,
    human_id: params.human_id,
    commissioning_authority: params.commissioning_authority,
    timestamp: new Date().toISOString(),
    purpose: params.purpose,
    initial_capabilities: params.initial_capabilities,
    risk_tier: params.risk_tier,
    principal_binding_reference: params.principal_binding_reference,
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Lifecycle, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Create a vitality report, hash-chained to the previous report.
 */
export async function createVitalityReport(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    agent_id: string;
    session_nonce: string;
    state: LifecycleState;
    metrics: VitalityMetrics;
    prev_report_hash: string;
  },
): Promise<VitalityReport> {
  const vitalityScore = computeVitalityScore(params.metrics);

  const payload = {
    dcp_version: '2.0' as const,
    agent_id: params.agent_id,
    session_nonce: params.session_nonce,
    timestamp: new Date().toISOString(),
    vitality_score: vitalityScore,
    state: params.state,
    metrics: params.metrics,
    prev_report_hash: params.prev_report_hash,
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Lifecycle, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Compute hash of a vitality report for chaining.
 */
export function hashVitalityReport(report: VitalityReport): string {
  const { composite_sig: _, ...payload } = report;
  return 'sha256:' + createHash('sha256').update(canonicalizeV2(payload)).digest('hex');
}

/**
 * Create a decommissioning record.
 */
export async function createDecommissioningRecord(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    agent_id: string;
    session_nonce: string;
    human_id: string;
    termination_mode: TerminationMode;
    reason: string;
    final_vitality_score: number;
    successor_agent_id: string | null;
    data_disposition: DataDisposition;
  },
): Promise<DecommissioningRecord> {
  const payload = {
    dcp_version: '2.0' as const,
    agent_id: params.agent_id,
    session_nonce: params.session_nonce,
    human_id: params.human_id,
    timestamp: new Date().toISOString(),
    termination_mode: params.termination_mode,
    reason: params.reason,
    final_vitality_score: params.final_vitality_score,
    successor_agent_id: params.successor_agent_id,
    data_disposition: params.data_disposition,
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Lifecycle, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

File: sdks/python/dcp_ai/v2/lifecycle.py
"""
DCP-05 v2.0 Agent Lifecycle Management — Python port.

State machine: commissioned -> active -> declining -> decommissioned.

Mirrors sdks/typescript/src/core/lifecycle.ts semantics exactly — every
artifact produced here is byte-identical to the TS output under the same
input (canonicalisation + composite sign over context "DCP-AI.v2.Lifecycle").
"""

from __future__ import annotations

import hashlib
from datetime import datetime, timezone
from typing import Any

from dcp_ai.v2.canonicalize import canonicalize_v2
from dcp_ai.v2.composite_ops import CompositeKeyInfo, composite_sign
from dcp_ai.v2.crypto_registry import AlgorithmRegistry
from dcp_ai.v2.domain_separation import DCP_CONTEXTS

# ── Lifecycle state machine ──

_LifecycleState = str  # "commissioned" | "active" | "declining" | "decommissioned"

_VALID_TRANSITIONS: dict[_LifecycleState, list[_LifecycleState]] = {
    "commissioned": ["active", "decommissioned"],
    "active": ["declining", "decommissioned"],
    "declining": ["decommissioned", "active"],
    "decommissioned": [],
}


def validate_state_transition(from_state: _LifecycleState, to_state: _LifecycleState) -> bool:
    """Return True iff the lifecycle transition from->to is allowed."""
    return to_state in _VALID_TRANSITIONS.get(from_state, [])


# ── Vitality scoring ──

_METRIC_WEIGHTS = {
    "task_completion_rate": 0.3,
    "error_rate": 0.25,
    "human_satisfaction": 0.25,
    "policy_alignment": 0.2,
}


def compute_vitality_score(metrics: dict[str, float]) -> int:
    """Compute a vitality score in 0..1000 from per-metric floats in 0..1.

    error_rate is inverted (lower error = higher score).
    """
    raw = (
        metrics["task_completion_rate"] * _METRIC_WEIGHTS["task_completion_rate"]
        + (1.0 - metrics["error_rate"]) * _METRIC_WEIGHTS["error_rate"]
        + metrics["human_satisfaction"] * _METRIC_WEIGHTS["human_satisfaction"]
        + metrics["policy_alignment"] * _METRIC_WEIGHTS["policy_alignment"]
    )
    clamped = max(0.0, min(1.0, raw))
    return round(clamped * 1000)


# ── Artifact creation ──


def _utc_now() -> str:
    # ISO-8601 UTC with Z suffix, matching the TS `new Date().toISOString()`.
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.") + (
        f"{datetime.now(timezone.utc).microsecond // 1000:03d}Z"
    )


async def create_commissioning_certificate(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    agent_id: str,
    session_nonce: str,
    human_id: str,
    commissioning_authority: str,
    purpose: str,
    initial_capabilities: list[str],
    risk_tier: str,
    principal_binding_reference: str,
) -> dict[str, Any]:
    """Create a commissioning certificate for a new agent."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "agent_id": agent_id,
        "session_nonce": session_nonce,
        "human_id": human_id,
        "commissioning_authority": commissioning_authority,
        "timestamp": _utc_now(),
        "purpose": purpose,
        "initial_capabilities": initial_capabilities,
        "risk_tier": risk_tier,
        "principal_binding_reference": principal_binding_reference,
    }

    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Lifecycle"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}


async def create_vitality_report(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    agent_id: str,
    session_nonce: str,
    state: _LifecycleState,
    metrics: dict[str, float],
    prev_report_hash: str,
) -> dict[str, Any]:
    """Create a vitality report hash-chained to the previous report."""
    vitality_score = compute_vitality_score(metrics)
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "agent_id": agent_id,
        "session_nonce": session_nonce,
        "timestamp": _utc_now(),
        "vitality_score": vitality_score,
        "state": state,
        "metrics": metrics,
        "prev_report_hash": prev_report_hash,
    }

    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Lifecycle"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}


def hash_vitality_report(report: dict[str, Any]) -> str:
    """Compute sha256 hash of a vitality report (composite_sig excluded) for chaining."""
    payload = {k: v for k, v in report.items() if k != "composite_sig"}
    canonical = canonicalize_v2(payload)
    return "sha256:" + hashlib.sha256(canonical.encode("utf-8")).hexdigest()


async def create_decommissioning_record(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    agent_id: str,
    session_nonce: str,
    human_id: str,
    termination_mode: str,
    reason: str,
    final_vitality_score: int,
    successor_agent_id: str | None,
    data_disposition: str,
) -> dict[str, Any]:
    """Create a decommissioning record."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "agent_id": agent_id,
        "session_nonce": session_nonce,
        "human_id": human_id,
        "timestamp": _utc_now(),
        "termination_mode": termination_mode,
        "reason": reason,
        "final_vitality_score": final_vitality_score,
        "successor_agent_id": successor_agent_id,
        "data_disposition": data_disposition,
    }

    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Lifecycle"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}
